node/silkworm/db/access_layer.cpp

```diff
@@ -190,29 +190,42 @@ void read_transactions(mdbx::cursor& txn_table, uint64_t base_id, uint64_t count
     SILKWORM_ASSERT(i == count);
 }
 
-bool read_block(mdbx::txn& txn, BlockNum block_number, bool read_senders, BlockWithHash& bh) {
-    // Locate canonical hash
+bool read_block(mdbx::txn& txn, BlockNum block_number, bool read_senders, Block& block) {
+    return read_blocks(txn, block_number, read_senders, {&block, 1}) == 1;
+}
 
+size_t read_blocks(mdbx::txn& txn, const BlockNum from, const bool read_senders, gsl::span<Block> out) {
+    Bytes key;
     Cursor canonical_hashes_cursor(txn, table::kCanonicalHashes);
-    auto key{block_key(block_number)};
-    auto data{canonical_hashes_cursor.find(to_slice(key), false)};
-    if (!data) {
-        return false;
-    }
 
-    SILKWORM_ASSERT(data.value.length() == kHashLength);
-    std::memcpy(bh.hash.bytes, data.value.data(), kHashLength);
+    for (size_t i{0}; i < out.size(); ++i) {
+        const BlockNum block_number{from + i};
+        Block& block{out[i]};
 
-    // Read header
-    key = block_key(block_number, bh.hash.bytes);
-    auto raw_header{read_header_raw(txn, key)};
-    if (raw_header.empty()) {
-        return false;
+        key = block_key(block_number);
+        auto data{canonical_hashes_cursor.find(to_slice(key), false)};
```

Comment:
Why build the key (which implies Bytes allocation and endianess swap) for each block number and force a find (which is slower than movenext) ?
kCanonicalHashes is already sorted by block number ... use a loop with a single find of initial key and then move next, move next till EOF or when count of retrieved elements reaches the limit.
Eventually endian load data key from cursor to check blocks are in order

Reply:
Fair enough, I've switched to `db::cursor_for_count`